### PR TITLE
Create fast TextControl::SetText variant.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR follows #1038.

### CODE REVIEW

File: GG/src/Font.cpp [hunk 1]
```diff
@@ -1253,140 +1438,143 @@ namespace DebugOutput {
...
+                element->text = Substring(text, (*it)[0]);
+
+                // Check open tags for submatches which are parameters.  For example a Color tag
+                // might have RGB parameters.
+                if (1 < (*it).nested_results().size()) {
+                    element->params.reserve((*it).nested_results().size() - 1);
+                    for (smatch::nested_results_type::const_iterator nested_it =
+                             ++(*it).nested_results().begin();
+                         nested_it != (*it).nested_results().end();
+                         ++nested_it) {
```
Comment (geoffthemedio) [inline GG/src/Font.cpp]: move { to next line

File: GG/src/Font.cpp [hunk 2]
```diff
@@ -1253,140 +1438,143 @@ namespace DebugOutput {
...
+    return text_elements;
+}
+
+
+void Font::FillTemplatedText(
+    const std::string& text,
+    std::vector<boost::shared_ptr<Font::TextElement> >& text_elements,
+    std::vector<boost::shared_ptr<Font::TextElement> >::iterator start) const
+{
+    std::vector<boost::shared_ptr<Font::TextElement> >::iterator& te_it = start;
```
Comment (geoffthemedio) [inline GG/src/Font.cpp:1579]: can you add some comments explaining what's going on in here?

File: GG/src/TextControl.cpp [hunk 3]
```diff
@@ -71,9 +71,10 @@ Pt TextControl::MinUsableSize(X width) const
     if ( m_cached_minusable_size_width != X0 &&  abs_delta_w < min_delta)
         return m_cached_minusable_size;
 
-    std::vector<Font::LineData> dummy_line_data;
     Flags<TextFormat> dummy_format(m_format);
-    m_cached_minusable_size = m_font->DetermineLines(m_text, dummy_format, width, dummy_line_data)
+    std::vector<Font::LineData> dummy_line_data =
```
Comment (geoffthemedio) [inline GG/src/TextControl.cpp:79]: ...

File: GG/src/TextControl.cpp [hunk 4]
```diff
@@ -183,9 +184,42 @@ void TextControl::SetText(const std::string& str)
...
+}
+
+void TextControl::SetText(const std::string& str,
+                          const std::vector<boost::shared_ptr<Font::TextElement> >& text_elements)
+{
+    if (!utf8::is_valid(str.begin(), str.end()))
+        return;
+    m_text = str;
+
+    if (!m_font)
```
Comment (geoffthemedio) [inline GG/src/TextControl.cpp]: does this test make sense here, after setting m_text but before setting m_code_points and m_text_elements?

File: GG/src/Font.cpp [hunk 5]
```diff
@@ -478,6 +484,180 @@ std::pair<std::size_t, CPSize> GG::LinePositionOf(CPSize index, const std::vecto
...
+        bool m_ignore_tags;
+
+        // m_tag_stack is used to track XML opening/closing tags.
+        std::stack<Font::Substring> m_tag_stack;
+
+        // The combined regular expression.
+        xpr::sregex m_EVERYTHING;
+    };
+
+    class TagHandler {
```
Comment (geoffthemedio) [inline GG/src/Font.cpp:597]: document class, at least briefly?

File: GG/src/Font.cpp [hunk 6]
```diff
@@ -478,6 +484,180 @@ std::pair<std::size_t, CPSize> GG::LinePositionOf(CPSize index, const std::vecto
...
+        // Return a regex bound to \p text using the currently known
+        // tags possible \p ignore_tags and/or \p strip_unpaired_tags
+        xpr::sregex & Regex(const std::string& text, bool ignore_tags, bool strip_unpaired_tags = false) {
+            if (!strip_unpaired_tags)
+                return m_regex_w_tags.BindRegexToText(text, ignore_tags);
+            else
+                return m_regex_w_tags_skipping_unmatched.BindRegexToText(text, ignore_tags);
+        }
+
+        private:
```
Comment (geoffthemedio) [inline GG/src/Font.cpp]: private: and public: should be de-indented from the following function / member declarations

File: GG/GG/Font.h [hunk 7]
```diff
@@ -147,6 +147,9 @@ class GG_API Font
             pair.second. */
         Substring(const std::string& str_, const IterPair& pair);
 
+        /** Rebind to a different string.*/
```
Comment (geoffthemedio) [inline GG/GG/Font.h]: Maybe already done in a different PR that includes the same additions, but some documentation / explanation of this / these function(s) would be helpful.

**Review by geoffthemedio — CHANGES_REQUESTED** [review]

File: GG/src/Font.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp:597]: Done in PR #1040.

File: GG/src/Font.cpp (hunk 6 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp]: Done in PR #1040.

File: GG/src/Font.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp]: Done in PR #1036.

File: GG/src/Font.cpp (hunk 2 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp:1579]: Done in #1039.

File: GG/src/TextControl.cpp (hunk 3 above)
Comment (LGM-Doyle) [reply GG/src/TextControl.cpp:79]: I was not exactly sure what "..." meant so I added comments.

File: GG/src/TextControl.cpp (hunk 4 above)
Comment (LGM-Doyle) [reply GG/src/TextControl.cpp]: I moved the test to just before m_font is used, here in #1040.

File: GG/GG/Font.h (hunk 7 above)
Comment (LGM-Doyle) [reply GG/GG/Font.h]: I expanded the documentation of these functions where they were defined in #1039.

**Comment by LGM-Doyle** [comment]
Rebased for ease of review.

File: GG/src/TextControl.cpp [hunk 8]
```diff
@@ -204,6 +204,40 @@ void TextControl::SetText(const std::string& str)
...
+    for (std::vector<boost::shared_ptr<Font::TextElement> >::iterator it = m_text_elements.begin();
+         it != m_text_elements.end(); ++it)
+    {
+        (*it)->Bind(m_text);
+    }
+
+    if (!m_font)
+        return;
+
+    m_line_data = m_font->DetermineLines(m_text, m_format, ClientSize().x, m_text_elements);
```
Comment (geoffthemedio) [inline GG/src/TextControl.cpp:226]: presumably a later change will avoid redoing this for multiple uses of the same text and text elements?

Comment (LGM-Doyle) [reply GG/src/TextControl.cpp:226]: Two things.  
1. PR #989 has a commit [Refactor TextControl::RecomputeLineData from SetText and ChangeTemplatedText](https://github.com/freeorion/freeorion/pull/989/commits/625341fd681ceebd74323f98f12573a037405239) , which refactors the repeated code into a function called `RecomputeLineData`.
2. This `DetermineLines()` function has had the slow parsing  code moved into `ExpensiveParseFromTextToTextElements()`.  It is only computing the `m_line_data, which means calculating where the line breaks are and adding the justification information to the line data so that prerender can place the glyphs in the line later.
